pkg/controller/core.oam.dev/v1alpha2/applicationrollout/rollout_handler.go

```diff
@@ -0,0 +1,236 @@
+}
+
+// extractWorkload use GVK and name of workload(assembled result) to fetch real workload in cluster
+func (h *rolloutHandler) extractWorkload(ctx context.Context, workload unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	wl, err := oamutil.GetObjectGivenGVKAndName(ctx, h, workload.GroupVersionKind(), workload.GetNamespace(), workload.GetName())
+	if err != nil {
+		return nil, err
+	}
+	return wl, nil
+}
+
+// if in middle of previous rollout, continue use previous source and target appRevision as this round rollout
+func (h *rolloutHandler) handleRolloutModified() {
+	klog.InfoS("rollout target changed, restart the rollout", "new source", h.appRollout.Spec.SourceAppRevisionName,
+		"new target", h.appRollout.Spec.TargetAppRevisionName)
+	h.record.Event(h.appRollout, event.Normal("Rollout Restarted",
+		"rollout target changed, restart the rollout", "new source", h.appRollout.Spec.SourceAppRevisionName,
+		"new target", h.appRollout.Spec.TargetAppRevisionName))
+	// we are okay to move directly to restart the rollout since we are at the terminal state
+	// however, we need to make sure we properly finalizing the existing rollout before restart if it's
+	// still in the middle of rolling out
+	if h.appRollout.Status.RollingState != v1alpha1.RolloutSucceedState &&
+		h.appRollout.Status.RollingState != v1alpha1.RolloutFailedState {
+		// happen when roll forward or revert in middle of rollout, previous rollout haven't finished
+		// continue to handle the previous resources until we are okay to move forward
+		h.targetRevName = h.appRollout.Status.LastUpgradedTargetAppRevision
+		h.sourceRevName = h.appRollout.Status.LastSourceAppRevision
+	} else {
+		// previous rollout have finished, go ahead using new source/target revision
+		h.targetRevName = h.appRollout.Spec.TargetAppRevisionName
+		h.sourceRevName = h.appRollout.Spec.SourceAppRevisionName
+		// mark so that we don't think we are modified again
+		h.appRollout.Status.LastUpgradedTargetAppRevision = h.appRollout.Spec.TargetAppRevisionName
+		h.appRollout.Status.LastSourceAppRevision = h.appRollout.Spec.SourceAppRevisionName
+	}
+	h.appRollout.Status.StateTransition(v1alpha1.RollingModifiedEvent)
+}
+
+// templateTargetManifest call dispatch to template target manifest to cluster
+func (h *rolloutHandler) templateTargetManifest(ctx context.Context) error {
+	var rt *v1beta1.ResourceTracker
+	// only when sourceAppRevision is not nil, we need gc old revision resources
+	if h.sourceAppRevision != nil {
+		rt := new(v1beta1.ResourceTracker)
+		err := h.Get(ctx, types.NamespacedName{Name: dispatch.ConstructResourceTrackerName(h.appRollout.Spec.SourceAppRevisionName, h.appRollout.Namespace)}, rt)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	// use source resourceTracker to handle same resource owner transfer
+	dispatcher := dispatch.NewAppManifestsDispatcher(h, h.targetAppRevision).EnableUpgradeAndSkipGC(rt)
```

Comment:
With help of `EnableUpgradeAndSkipGC`, this PR is expected to fix #1728 . I suggest you add an e2e-test for it.